core/base/batch_multi_vector.cpp

```diff
@@ -0,0 +1,256 @@
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/base/batch_multi_vector.hpp>
+
+
+#include <algorithm>
+#include <type_traits>
+
+
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/exception.hpp>
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/base/matrix_data.hpp>
+#include <ginkgo/core/base/utils.hpp>
+
+
+#include "core/base/batch_multi_vector_kernels.hpp"
+
+
+namespace gko {
+namespace batch_multi_vector {
+
+
+GKO_REGISTER_OPERATION(scale, batch_multi_vector::scale);
+GKO_REGISTER_OPERATION(add_scaled, batch_multi_vector::add_scaled);
+GKO_REGISTER_OPERATION(compute_dot, batch_multi_vector::compute_dot);
+GKO_REGISTER_OPERATION(compute_norm2, batch_multi_vector::compute_norm2);
+GKO_REGISTER_OPERATION(copy, batch_multi_vector::copy);
+
+
+}  // namespace batch_multi_vector
+
+
+template <typename ValueType>
+void BatchMultiVector<ValueType>::scale_impl(
+    const BatchMultiVector<ValueType>* alpha)
+{
+    GKO_ASSERT_EQ(alpha->get_num_batch_entries(),
+                  this->get_num_batch_entries());
+    GKO_ASSERT_EQUAL_ROWS(alpha->get_common_size(), dim<2>(1, 1));
+    for (size_type b = 0; b < alpha->get_num_batch_entries(); ++b) {
+        if (alpha->get_common_size()[1] != 1) {
+            // different alpha for each column
+            GKO_ASSERT_EQUAL_COLS(this->get_common_size(),
+                                  alpha->get_common_size());
+        }
+    }
+    this->get_executor()->run(batch_multi_vector::make_scale(alpha, this));
+}
+
+
+template <typename ValueType>
+void BatchMultiVector<ValueType>::add_scaled_impl(
+    const BatchMultiVector<ValueType>* alpha,
+    const BatchMultiVector<ValueType>* b)
+{
+    GKO_ASSERT_EQ(alpha->get_num_batch_entries(),
+                  this->get_num_batch_entries());
+    GKO_ASSERT_EQUAL_ROWS(alpha->get_common_size(), dim<2>(1, 1));
+    for (size_type b = 0; b < alpha->get_num_batch_entries(); ++b) {
+        if (alpha->get_common_size()[1] != 1) {
+            // different alpha for each column
+            GKO_ASSERT_EQUAL_COLS(this->get_common_size(),
+                                  alpha->get_common_size());
+        }
+    }
+    GKO_ASSERT_EQ(b->get_num_batch_entries(), this->get_num_batch_entries());
+    GKO_ASSERT_EQUAL_DIMENSIONS(this->get_common_size(), b->get_common_size());
```

Comment:
Perhaps an assertion `GKO_ASSERT_EQUAL_BATCHES` that combines the two would help.

Reply:
I do have that in the batch_develop branch, but there we also have more complex checks as we support non-uniform batches. I think for these simple ones, I feel an additional macro is probably unnecessary.

Reply:
I would still prefer if that macro would exist. Remembering to use one macro is simpler than remembering to use two different ones. But I think that can be handled in a different PR.